### PR TITLE
Improve Parallel Processing and Add New Benchmark Feature

### PR DESCRIPTION
**Pull request by kmati-dev**
### Improvements
**Parallel Processing**: Changed from ```multiprocessing``` to ```concurrent.futures``` (Threading)
**Rationale**: Multiprocessing is not optimal for I/O tasks. Switching to threading enhances I/O task handling.
**Benefit**: This improvement allows us to handle more than 64 concurrent requests on an M1 (as Client).

### New Feature
**Benchmarking with OpenAPI Endpoint Compatibility:**
Added support for benchmarking via ```base_url``` and ```api_key```.

### Additional Information
**Benchmark Results:**
Tested ```scb10x/llama-3-typhoon-v1.5x-8b-instruct``` and ```xl_sum_tha_seacrowd_t2t``` with L4 1 card.
For 826 samples, the process took approximately 26 minutes.

Cheers 🎉🎈🎊
Mati

### CODE REVIEW

File: runner.sh [hunk 1]
```diff
@@ -5,4 +5,13 @@ python evaluation/main_nlg_prompt_batch.py tha $MODEL_NAME 0 4
 python evaluation/main_llm_judge_batch.py $MODEL_NAME --data ThaiLLM-Leaderboard/mt-bench-thai
 python scripts/transform_result.py $MODEL_NAME
 # clear model cache
-rm -rf ~/.cache/huggingface/hub/
\ No newline at end of file
+rm -rf ~/.cache/huggingface/hub/
+
+# run with OpenAI API compatible
```
Comment (kunato) [inline runner.sh:10]: Please move this section into new file and specific that on README.md  eval_openai_compatible.sh

File: evaluation/model_utils.py [hunk 2]
```diff
@@ -308,15 +318,19 @@ def predict_generation(
...
-
-        with Pool(min(8, len(prompts))) as p:
-            _generate_fn = partial(
-                self.generate_fn,
-                model_name=self.model_name,
-                max_tokens=self.max_generation_length,
-                **kwargs
-            )
-            results = p.map(_generate_fn, prompts)
+        # with Pool(min(8, len(prompts))) as p:
```
Comment (kunato) [inline evaluation/model_utils.py:321]: Remove commented code and leave only new code

File: evaluation/model_utils.py [hunk 3]
```diff
@@ -278,9 +285,13 @@ def predict_classification(
             for prompt in prompts
         ]
         hyps = []
-        with Pool(min(8, len(prompts))) as p:
-            _generate_fn = partial(self.generate_fn, model_name=self.model_name)
-            results = p.map(_generate_fn, inputs)
+        # with Pool(min(8, len(prompts))) as p:
```
Comment (kunato) [inline evaluation/model_utils.py:288]: Remove commented code and leave only new code

File: evaluation/main_nlu_prompt_batch.py [hunk 4]
```diff
@@ -75,7 +84,9 @@ def to_prompt(input, prompt, labels, prompt_lang, schema):
     # Set seed before initializing model.
     set_seed(42)
 
-    model_runner = load_model_runner(MODEL)
+    # model_runner = load_model_runner(MODEL)
```
Comment (kunato) [inline evaluation/main_nlu_prompt_batch.py:87]: Remove commented code and leave only new code

File: runner.sh (hunk 1 above)
Comment (kunato) [reply runner.sh:10]: also please add information that due to openai compatible cannot access logits, the result on nlu task is not directly comparable to the official leaderboard

File: evaluation/main_nlu_prompt_batch.py (hunk 4 above)
Comment (kunato) [reply evaluation/main_nlu_prompt_batch.py:87]: Thank you for your contribution! Let's work on adding this new feature to the project.

Here are my comments:
1. Please clean up the existing comments in the code.
2. Move the run instructions to a separate file.
3. Add instructions on how to run the project on an OpenAI-compatible server in the README. (Also, please include a note that since OpenAI-compatible servers cannot access logits, the results on NLU tasks may not be directly comparable to the official leaderboard.)

Comment (matichon-vultureprime) [reply evaluation/main_nlu_prompt_batch.py:87]: Regarding the third topic, fortunately, 
TensorRT-LLM can respond with both context logits and generation logits.

So, I will validate the result before committing again.
